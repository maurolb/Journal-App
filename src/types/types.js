export const types = {
  login: '[Auth] Login',
  logout: '[Auth] Logout',

  uiSetError: '[UI] Set Error',
  uiRemoveError: '[UI] Remove Error',

  uiStartLoading: '[UI] Start Loading',
  uiFinishLoading: '[UI] Finish Loading',

  notesAddNew: '[Notes] New Note',
  notesActive: '[Notes] Set Active Note',
  notesLoad: '[Notes] Load Notes',
  notesUpdate: '[Notes] Update Note Saved',
  notesDelete: '[Notes] Delete Note',
  notesLogoutCleaning: '[Notes] Logout Cleaning',
};