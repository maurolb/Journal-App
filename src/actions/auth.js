import Swal from 'sweetalert2';
import { getAuth, updateProfile ,signInWithPopup, createUserWithEmailAndPassword, signInWithEmailAndPassword } from 'firebase/auth';
import { googleProvider } from '../firebase/firebase-config';
import { types } from "../types/types";
import { finishLoading, startLoading } from './ui';
import { noteLogout } from './notes';

export const startLoginEmailPassword = ( email, password ) => {
  return ( dispatch ) => {

    dispatch( startLoading() );

    const auth = getAuth();
    return signInWithEmailAndPassword( auth, email, password )
      .then( ({ user }) => {
        dispatch( login( user.uid, user.displayName ) );
        dispatch( finishLoading() );
      })
      .catch(( err ) => {
        console.log( err );
        dispatch( finishLoading() );
        Swal.fire('Error', err.message, 'error');
      });
  };
};

export const startRegisterWithEmailPasswordName = ( email, password, name ) => {
  return ( dispatch ) => {
    const auth = getAuth();
    createUserWithEmailAndPassword( auth, email, password )
      .then( async ({ user }) => {
        await updateProfile( user, {displayName: name} );
        dispatch( login( user.uid, user.displayName ) );
      })
      .catch(( err ) => {
        console.log( err );
        Swal.fire('Error', err.message, 'error');
      });
      
  };
};

export const startGoogleLogin = () => {
  return ( dispatch ) => {
    const auth = getAuth();
    signInWithPopup( auth, googleProvider )
      .then(({ user }) => {
          dispatch( login( user.uid, user.displayName ) );
      });
  };
};

export const login = ( uid, displayName ) => {
  return {
    type: types.login,
    payload: {
      uid,
      displayName
    }
  };
};

export const startLogout = () => {
  return async ( dispatch ) => {
    await getAuth().signOut();
    dispatch( logout() );
    dispatch( noteLogout() );
  };
};

export const logout = () => ({
  type: types.logout
});