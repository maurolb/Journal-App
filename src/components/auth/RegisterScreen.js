import React from 'react';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import validator from 'validator';
import { useDispatch, useSelector } from 'react-redux';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen = () => {

  const dispatch = useDispatch();
  const { msgError } = useSelector( state => state.ui );

  const [ formValues, handleInputChange ] = useForm({});

  const { name, email, password, confirm } = formValues;

  const handleRegister = (e) => {
    e.preventDefault();
    
    if( isFormValid() ) {
      dispatch( startRegisterWithEmailPasswordName( email, password, name ) );
    };

  };

  const isFormValid = () => {
    if( name.trim().length === 0 ){
      dispatch( setError('Name is required') );
      return false;
    } else if( !validator.isEmail( email ) ) {
      dispatch( setError('Email is not valid') );
      return false;
    } else if( password !== confirm || password < 5 ) {
      dispatch( setError('assword should be a least 6 characters and match each other') );
      return false;
    }

    dispatch( removeError() );

    return true;
  };

  return (
    <>
      <h3 className="auth__title">Register</h3>

      <form
        onSubmit={ handleRegister }
        className="animate__animated animate__fadeIn animate__faster"
      >

        {
          msgError && (
            <div className="auth__alert-error">
              { msgError }
            </div>
          )
        }

        <input
          type="text"
          placeholder="Name"
          name="name"
          className="auth__input"
          autoComplete="off"
          value={ name }
          onChange={ handleInputChange }
        />

        <input
          type="text"
          placeholder="Email"
          name="email"
          className="auth__input"
          autoComplete="off"
          value={ email }
          onChange={ handleInputChange }
        />

        <input
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
          value={ password }
          onChange={ handleInputChange }
        />
        
        <input
          type="password"
          placeholder="Confirm password"
          name="confirm"
          className="auth__input"
          value={ confirm }
          onChange={ handleInputChange }
        />

        <button
          type="submit"
          className="btn btn-primary btn-block mb-5"
        >
          Register
        </button>
        
        <Link to="/auth/login" className="link">Already register?</Link>
      </form>
    </>
  );
};
