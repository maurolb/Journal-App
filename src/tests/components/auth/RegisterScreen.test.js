import configureStore from 'redux-mock-store';
import thunk from "redux-thunk"; 
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { RegisterScreen } from "../../../components/auth/RegisterScreen";
import { types } from '../../../types/types';

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initialState = {
  auth: {},
  ui: {
    loading: false,
    msgError: null
  }
}

let store = mockStore( initialState );

const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <RegisterScreen />
    </MemoryRouter>
  </Provider>
);

describe('Tests on RegisterScreen component', () => {
  
  // beforeEach( () => {
  //   store = mockStore( initialState );
  //   jest.clearAllMocks();
  // });

  test('should show correctly', () => {

    expect( wrapper ).toMatchSnapshot();

  });
  

  test('should asd', () => {
    
    const emailField = wrapper.find('input[name="email"]'); // seleccion por name

    emailField.simulate('change', {
      target: {
        value: '',
        name: 'email'
      }
    });

    wrapper.find('form').simulate('submit', {
      preventDefault(){}
    });

    const actions = store.getActions();
    
    expect( actions[0] ).toEqual({
      type: types.uiSetError,
      payload: 'Email is not valid'
    });
  });
  

  test('should show error box', () => {
    
    const initialState = {
      auth: {},
      ui: {
        loading: false,
        msgError: 'Email incorrecto'
      }
    }
    
    const store = mockStore( initialState );
    
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <RegisterScreen />
        </MemoryRouter>
      </Provider>
    );
    
    expect( wrapper.find('.auth__alert-error').exists() ).toBe( true );
    expect( wrapper.find('.auth__alert-error').text().trim() ).toBe( initialState.ui.msgError );

  });
  

});
