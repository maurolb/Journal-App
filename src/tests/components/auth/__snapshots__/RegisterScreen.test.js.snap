// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on RegisterScreen component should show correctly 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="mauro"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="mauro@mauro.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="asd123"
    />
    <input
      className="auth__input"
      name="confirm"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="asd123"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already register?
    </a>
  </form>,
]
`;
